component/stream/stream_cache.go

```diff
@@ -80,104 +80,61 @@ type cache struct {
-	c.Lock()
-	defer c.Unlock()
-	f := c.files[fileKey]
-	f.openHandles -= 1
-	return f.openHandles
-}
-
-// on file opens we increment handles for a given file
-func (c *cache) incrementHandles(fileKey string) {
-	c.Lock()
-	defer c.Unlock()
-	f := c.files[fileKey]
-	f.openHandles += 1
-}
-
 // add a new file key and create a cache object for it to hold references to its current stored blocks
-func (c *cache) addFileKey(fileKey string) {
-	c.Lock()
-	defer c.Unlock()
-	_, ok := c.files[fileKey]
-	if !ok {
-		log.Trace("streamcache:: file key %s not found initializing new key", fileKey)
-		//EvictedFunc is a callback that allows us to make sure we remove the corresponding blocks from the block cache
-		// PurgeVisitorFunc is a callback that allows us to walk through each block cache entry as we're purging it and perform a cleanup operation we want
-		var fc gcache.Cache
-		switch c.evictionPolicy {
-		case common.EPolicy.LFU():
-			fc = gcache.New(c.blocksPerFileKey).LRU().EvictedFunc(c.fileEvict).PurgeVisitorFunc(c.filePurge).Build()
-		case common.EPolicy.ARC():
-			fc = gcache.New(c.blocksPerFileKey).LFU().EvictedFunc(c.fileEvict).PurgeVisitorFunc(c.filePurge).Build()
-		default:
-			fc = gcache.New(c.blocksPerFileKey).ARC().EvictedFunc(c.fileEvict).PurgeVisitorFunc(c.filePurge).Build()
-		}
-		c.files[fileKey] = &cacheFile{
-			fileBlockBuffer: fc,
-		}
+func (c *cache) addHandleCache(handle *handlemap.Handle) {
+	//EvictedFunc is a callback that allows us to make sure we remove the corresponding blocks from the block cache
+	// PurgeVisitorFunc is a callback that allows us to walk through each block cache entry as we're purging it and perform a cleanup operation we want
+	var fc gcache.Cache
+	switch c.evictionPolicy {
+	case common.EPolicy.LFU():
+		fc = gcache.New(c.blocksPerFileKey).LRU().EvictedFunc(c.fileEvict).PurgeVisitorFunc(c.filePurge).Build()
+	case common.EPolicy.ARC():
+		fc = gcache.New(c.blocksPerFileKey).LFU().EvictedFunc(c.fileEvict).PurgeVisitorFunc(c.filePurge).Build()
+	default:
+		fc = gcache.New(c.blocksPerFileKey).ARC().EvictedFunc(c.fileEvict).PurgeVisitorFunc(c.filePurge).Build()
 	}
-}
-
-func (c *cache) removeFileKey(fileKey string) {
-	c.Lock()
-	defer c.Unlock()
-	// remove all blocks stored for the file key
-	// the purge walker callback will happen which will walk through all the block refs for this file and remove it from the main block cache
-	c.files[fileKey].fileBlockBuffer.Purge()
-	// delete the entry from our map once complete
-	delete(c.files, fileKey)
+	handle.DataBuffer = fc
 }
 
 // try to retrieve the block - return missing if it is not cached
-func (c *cache) getBlock(fileKey string, offset int64, fileSize int64) (*cacheBlock, bool) {
+func (c *cache) getBlock(handle *handlemap.Handle, offset int64) (*cacheBlock, bool) {
 	blockSize := c.blockSize
-	blockKeyObj := blockKey{fileKey: fileKey, offset: offset}
+	blockKeyObj := blockKey{handle: handle, offset: offset}
 	c.Lock()
```

Comment:
Do we still need this lock? as blocks are within the handle itself policy level locks are no more needed I guess.

Reply:
Its not possible to remove those locks :/ it would cause a deadlock in some cases if eviction needs to happen on the same handle that we're adding a new block to. And the solution around that deadlock would mean the possibility of duplicate blocks in the buffer.

Reply:
we can have a handle level lock and a block level lock. So before any addition/deletion of block lock the handle -> lock thta particular block and then do the operation. For regular read-write lock the handle, get the block, lock the block, unlock the handle. this shall not create any deadlock I suppose.